### DISCUSSION TITLE
YOURLs 1.8.2 Subdirectory Setup

### DISCUSSION THREAD

**Discussion by golddave** [YOURLS/YOURLS#3081] [General]
I have been using YOURLS since 2010. I have been running it from a subdirectory but creating links from the root by using the method described here: http://nothing.golddave.com/2010/12/16/generating-root-short-urls-from-a-yourls-subdirectory-install/ and updated for 1.7.9 here: https://nothing.golddave.com/2020/08/28/using-yourls-in-a-subdirectory-revisited/ This has been working flawlessly through 1.7.9.

Today I installed 1.8.2 and it doesn't work. I did some research and came upon this article: https://github.com/YOURLS/YOURLS/wiki/Giving-YOURLS-its-own-directory Using this method I'm getting short URLs that include the subdirectory in them and whenever I go to one (even an older one that existed before the upgrade) I get a 404 error. Does anyone have any experience with this in 1.8.2 and/or can help me troubleshoot?

Thanks.

**Comment by ozh** [comment]
In your `config.php` what is `YOURLS_SITE` set to ?

**Reply by golddave** [reply]
http://domain.com/yourls_subdirectory
Whene "domain.com" in my domain name and "yourls_subdirectory" is the directory the YOURLs are in.

**Reply by ozh** [reply]
Not sure what you're trying to do here.

The guide in the wiki is for people who installed YOURLS somewhere, but don't want all the YOURLS files in that directory.
So, you install YOURLS, say in http://sho.rt/ or http://mysite.com/yourls, and then move YOURLS files (yourls-api.php, yourls-go.php, etc...) into a subdirectory, like http://sho.rt/the_yourls_files_here or http://mysite.com/yourls/the_yourls_files_here . Hence, `YOURLS_SITE` points to where YOURLS is supposedly installed and serving short URL, ie http://sho.rt/, not http://sho.rt/the_yourls_files_here

**Reply by golddave** [reply]
I want the files to reside in http://domain.com/yourls_subdirectory and the short URLs to come from the root (http://domain.com/). I've had it working this way for years until now using the method described in the links from my earlier post but that method seems to no longer work in this version so I followed the wiki instructions and here I am.

**Reply by ozh** [reply]
OK so make `YOURLS_SITE` be `http://domain.com/` ?

**Reply by golddave** [reply]
I've tried that (both with and without the trailing /) and still get a 404 when I try to use a URL.

I just tried it again with a fresh install of the files (in case my earlier attempts muddled things) and I'm not seeing any of the images in the admin and none of the admin functionality is working.

**Reply by golddave** [reply]
Update: With the solution in the wiki not working for me, I made another small code change modifying the yourls_link function in functions-links.php  to change `yourls_get_yourls_site()` to `substr(yourls_get_yourls_site(), 0, 19) `, This changed the display of links in the admin to show short URLs without the `/yourls` subdirectory in them (which is what I want). However, the short links themselves get a 403 (not a 404) when used.

(I know this is a hack-y way of doing it and wish it would work for me another way but the wiki way didn't work so I'm stuck with this. I'm definitely open to other suggestions.)

**Reply by golddave** [reply]
It appears that the short URL is actually forwarding to the subdirectory then getting a 403. For example, http://golddave.com/ozhblog is giving a 403 on http://golddave.com/yourls/ instead of forwarding to https://planetozh.com/blog/. This happens whether I'm doing my hacky way or if I do it by the wiki. 

http://golddave.com/yourls/ = where the YOURLs files reside
http://golddave.com/?????? = the format of the short URLs where ?????? is the keyword
This was working with a hack up to version 1.7.9 but not working in 1.8.x

**Reply by ozh** [reply]
Works for me : 

* YOURLS installed at test.yourls.org. 
* I moved EVERYTHING into `/the_yourls_files_here`
![image](https://user-images.githubusercontent.com/223647/136656640-b335fc99-f5e9-41e0-a67e-3764c2b63179.png)
* the .htaccess file in root is just as in the wiki - the .htaccess the subdir is edited to reflect the directory (kind of obvious step I thought but indeed omitted from the wiki page)
![image](https://user-images.githubusercontent.com/223647/136656696-a7770257-5377-4cbc-abb5-dda9aafd4357.png)
* the config file is unchanged 
![image](https://user-images.githubusercontent.com/223647/136656829-5a8ffc78-7b47-40a9-8fbb-11a72e243c8a.png)
* https://test.yourls.org/admin/index.php works
* Creating and redirecting works : https://test.yourls.org/uwyf
* no hack, editing core files or plugin used
* I guess there could be some additional simplification of the .htaccess files, but well, it works, so I'm not trying to fix what's working

**Reply by golddave** [reply]
The difference is that you're using a subdomain. I'm not using a subdomain and don't want my short URLs to include a subdomain (if I were to use one). I'm also using the domain for other things that I don't want interrupted but if I use that .httaccess all traffic is redirected to the YOURLs admin.

**Reply by ozh** [reply]
Domain, subdomain, root or directory, that doesn't change a thing (Unless your hosting is really fucked up or has a convoluted setup maybe? On mine, ~/site.tld or ~/sub.site.tld are just two directories off my home directory)

And no matter what, YOURLS needs its own .htaccess

**Reply by golddave** [reply]
I tried this again following your steps and YOURLs itself seems to be working ok now. However, when I browse to the root of my domain (golddave.com) my browser is giving me an error that there are too many redirects instead of the page at the root of my domain. Further, when I go to other pages/directories on my domain they are loading incorrectly (in some cases formatting is missing, in other some functionality isn't working). There are some other issue as well. I believe this is the root .htaccess wreaking havoc on the rest of the site.

**Reply by ozh** [reply]
Indeed, has the same problem when trying to use a custom index file. Every use case has to be fine tuned to match need. In my case, fixed with: 

* at the root : an `index.php`
* updated root .htaccess : `RewriteRule ^(/index.php)?$ /index.php [L]` right after the line with `RewriteEngine On` 
![image](https://user-images.githubusercontent.com/223647/136700740-e2084b9c-f227-4595-aab8-be417e70b8ec.png)

Now : 
* root page now [works](https://test.yourls.org/)
* admin area still [works](https://test.yourls.org/admin/)
* creating and redirecting also still [work](https://test.yourls.org/nl0h)

Again, moving YOURLS files to their own subdir while still serving from current dir has to be fine tuned to match custom needs. And the limitation that YOURLS still needs its own .htaccess (as does any PHP app that handles URL rewrites) still exists.

**Reply by golddave** [reply]
Am I going to have to add a rewrite rule for every other page/subdirectory on the server?

**Reply by ozh** [reply]
I don't know. It depends on your setup, what you're trying to do, what's installed besides YOURLS and what it needs. Maybe your use case is different than what's described in the wiki and detailed here. Maybe a plugin is a better solution in your case. Maybe what your trying to achieve is a bad idea. I don't know :)

**Reply by golddave** [reply]
Thanks. I appreciate the help. I'm going to run with this for a while and see if I run into any issues.
